### DISCUSSION TITLE
Is there anyway to manage permission for users?

### DISCUSSION THREAD

**Discussion by Braggedtooth** [Leantime/leantime#857] [General] [answered]
I noticed the default permissions are not optimal  can we change the permissions per project and per users?

**Comment by marcelfolaron** [comment]
That is currently not possible but in the works. Would you mind sharing how
you would like to change the permissions?
I see per project which makes sense. What type of roles?

**Reply by Braggedtooth** [reply]
By  'per user' i mean setting different permissions for users under a project. 
it seems to me that all users have the same permissions when assigned to a project. 
would be nice to have some control over this. 
basic permission over crud actions. most importantly delete and update

**Reply by nuukth** [reply]
Exactly this. Currently, this is the main reason I am not using Leantime. It would be a perfect system for managing projects with a customer/client - but I don't want the customer to be able to move tasks around on the board (changing status). I don't want the user to be able to delete/update tasks when they are no longer in the backlog etc.

So, for me, I would really like something like this:

1. Create user/role with access to specific project.
2. Have a backlog with permission set, e.g. allow user/role to create new To-Do's, and edit/delete them as long as they are in the backlog/have status "New".
4. The second a To-Do is no longer in the backlog/have status "New", the user/role can no longer update/delete, but only add comments etc.


Adding CRUD permissions throughout the system, on individual functionality would be nice. But this Business/Client relationship would require that they could participate in adding To-Do's (request features/support/whatever), but when I (the Business) removes the To-Do from the Backlog ("New" status), i have to be the one in charge of that task, since it is now in my work-flow. The client should not, without any reason, be able to update or delete a To-Do. That's what comments can be used for.

**Comment by marcelfolaron** [comment, marked as answer]
I spend some time mapping out the roles. Here are my thoughts, open for feedback:

- Making roles entirely custom is going to be incredibly difficult at this stage in leantime
- Instead I'd like to make sure we cover 95% of use cases with a static role system
- All roles could be assigned at instance and project level (with project level roles overriding instance roles (except admins)

I made a spreadsheet: https://docs.google.com/spreadsheets/d/1LxsjAlXp9NAKS319OQkX_d-rcN-cbkSGbmZNyoFjYr8/edit?usp=sharing 
(feel free to comment). Otherwise, the data is below: 

<google-sheets-html-origin><style type="text/css"></style>

Feature | Owner | Admin | Manager | Editor | Commenter | Read-Only | External/Client
-- | -- | -- | -- | -- | -- | -- | --
  |   |   |   |   |   |   |  
**ToDos** |   |   |   |   |   |   |  
View Todos | x | x | x | x | x | x | x
Create | x | x | x | x |   |   | x
Edit Todos | x | x | x | x |   |   |  
Delete Todos | x | x | x | x |   |   |  
Comment | x | x | x | x | x |   | x
File Upload | x | x | x | x | x |   | x
  |   |   |   |   |   |   |  
**Ideas** |   |   |   |   |   |   |  
View | x | x | x | x | x | x | x
Create | x | x | x | x |   |   | x
Edit | x | x | x | x |   |   |  
Delete | x | x | x | x |   |   |  
Comment | x | x | x | x | x |   | x
File Upload | x | x | x | x | x |   | x
  |   |   |   |   |   |   |  
**Research** |   |   |   |   |   |   |  
View | x | x | x | x | x | x | x
Create | x | x | x | x |   |   | x
Edit | x | x | x | x |   |   |  
Delete | x | x | x | x |   |   |  
Comment | x | x | x | x | x |   | x
File Upload | x | x | x | x | x |   | x
  |   |   |   |   |   |   |  
**Timesheets** |   |   |   |   |   |   |  
View | x | x | x | x | x |   | x
Create | x | x | x | x |   |   | x
Edit | x | x | x | x |   |   |  
Delete | x | x | x | x |   |   |  
Comment | x | x | x | x | x |   | x
File Upload | x | x | x | x | x |   | x
  |   |   |   |   |   |   |  
**Projects** |   |   |   |   |   |   |  
View | x | x | x | x |   |   | x
Create | x | x | x |   | x | x |  
Edit | x | x | x |   |   |   |  
Delete | x | x | x |   |   |   |  
Comment | x | x | x |   |   |   | x
File Upload | x | x | x | x |   |   | x
Assign | x | x | x | x |   |   |  
  |   |   |   |   |   |   |  
**Users** |   |   |   |   |   |   |  
View Profiles | x | x | x | x | x | x | x
View All | x | x |   |   |   |   |  
Create | x | x |   |   |   |   |  
Edit | x | x |   |   |   |   |  
Delete | x | x |   |   |   |   |  
Comment | x | x | x | x |   |   | x
File Upload | x | x | x | x |   |   | x
Assign to project | x | x | x |   |   |   |  
  |   |   |   |   |   |   |  
**Retrospectives** |   |   |   |   |   |   |  
View | x | x | x | x | x | x | x
Create | x | x | x | x |   |   | x
Edit | x | x | x | x |   |   |  
Delete | x | x | x | x | x |   |  
Comment | x | x | x | x | x |   | x
File Upload |   |   |   |   |   |   | x
  |   |   |   |   |   |   |  
**Client** |   |   |   |   |   |   |  
View | x | x | x | x |   |   | x (own)
Create | x | x |   |   |   |   |  
Edit | x | x |   |   |   |   |  
Delete | x | x |   |   |   |   |  
Comment | x | x | x | x |   |   | x
File Upload | x | x | x | x |   |   | x
Assign | x | x | x |   |   |   |  
  |   |   |   |   |   |   |  
**Company Settings** |   |   |   |   |   |   |  
View | x | x |   |   |   |   |  
Edit | x |   |   |   |   |   |  
  |   |   |   |   |   |   |

**Reply by Braggedtooth** [reply]
I think this looks like a nice start.  
an idea could be to add an additional role that is customizable per project?

**Reply by nuukth** [reply]
It would definitely cover my use cases.
I've been looking for a system like this for SO long now - and I've tried Leantime a couple of times, but this permission thing stopped me every single time.

I am pretty convinced that with the permissions you detailed above, I would finally be able to go all-in on Leantime.

**Reply by diderich** [reply]
From my perspective, in addition to having role-based access control, there should be a possibility to have object based access control.

Some use cases:
- A user wants to create a personal ideation board where only he/she has access to.
- A project manager wants, at one point, to disable changes to a leancanvas (to freeze a developed version), and only allow comments.
- A user wants to offer read-only (and commenting) rights to their research board to other members of the project team so that they can have a look at it, comment, but cannot change the actual content.
- A user wants to add a specific task for himself to the project that is not relevant to other users (and therefore should not appear on their to-do board)

**Reply by marcelfolaron** [reply]
@diderich I agree, that would be a great addition. So far I haven't had a chance to really dive deep into this. May be something for a future release.

**Reply by diderich** [reply]
I don't think it would be very difficult to implement (one could mimic somehow the ACL concept of Linux in a simplified way), But I agree, that the system needs to settle with all the currently already planned changes, before adding a feature that will touch all areas of the system!

**Comment by Community-Mailex** [comment]
I have a problem, I would like to involve my clients in my project but I need to be able to trace their actions. What do you think about :

- Disable the ability to delete and edit comments.
- Create a role that can create tasks but cannot modify those of others.

**Reply by phulstaert** [reply]
Isn't that exactly what the "External/Client" is?

**Comment by Community-Mailex** [comment]
![GIT](https://user-images.githubusercontent.com/109607773/186116501-50aca622-8400-4bd6-9740-6d2005429c76.png)

This totally answers the second point. However, I don't have access to it from my interface, despite the last update.

**Reply by phulstaert** [reply]
The clients are made elsewhere.

![image](https://user-images.githubusercontent.com/8971696/186119235-c9a1406b-b3c7-4fbd-9eb9-d8ef3e5d48f6.png)

**Reply by phulstaert** [reply]
Never mind my remark, I see what you mean :-)

**Reply by phulstaert** [reply]
It appears this is still in the design phase. (I must admit I never let clients login in any system - maybe only a service desk system)

**Reply by Community-Mailex** [reply]
Ok, so if I understand correctly, we can hope to have this option in one of the next updates?

**Reply by phulstaert** [reply]
There is no guarantee that a feature request is completed for the next release.  It depends on when someone wants to work on it. 
You are most welcome to help, if you have the knowledge and time for it.

**Reply by Community-Mailex** [reply]
Ok great, I'll ask our developer if he can help with that! We've been using LeanTime for a short time, it's changed the way we work, it could be a good investment.

**Comment by marcelfolaron** [comment]
Hi,

the philosophy behind the current role system has been that external
clients would either just see whats going on or comment in existing tags.
If you would like to collaborate they should be editors.

With that said I think the ability to create tasks (but not necessarily
collaborate) is more of a help desk/ticketing use case.
we have been exploring the option to create a Todo intake form that can be
shared publicly.

I can add that as a feature request and we can discuss details in there.

**Reply by nuukth** [reply]
Just from my use case, create tasks and commenting is basically what I would need. Help Desk systems are also crap out there - it's either something so simple that it's not worth the time, or so complicated/complex (and expensive) that it doesn't make sense for a smaller business.

LeanTime is somewhere in between. It's simple and not extremely complex, but still has the functionality needed.... And also doesn't look like something where the UI got abandoned in 2005. ;)

Having the option for a client to create a new To-Do, and from there on only comment on it, would for me be a good way for a client to be an active part of breaking down the tasks what has to be done, with descriptions, priorities etc, but besides that only be able to keep track on progress, and add comments.

A ticketing system is more like a "You have an issue or a request for a new feature, and now we communicate through this tool, and you can keep track on when I expect to resolve your issue or create that new feature that you wanted".

I mean.. If a client is able to create To-Do's, I guess you could use the system for that purpose as well. But looking at other systems like Zammad, which is a great Open Source tool, but very complex to setup and administer (especially with multiple clients), and extremely resource demanding, I could see LeanTime as a good replacement if you just need basic "Help Desk/ticketing"-functionality. And I don't think it's necessarily a bad thing, and I don't think we're crossing that line of "Now we're trying to do 2 things at the same time, and fail at both".

Just my input here. :)

**Comment by marcelfolaron** [comment]
Yeah, usually if we have enough people asking for it, we’ll try to get it
in within the next two releases.

**Comment by marcelfolaron** [comment]
Awesome! Happy to jump on a call and help out where I can.
Also, you can join our discord server for faster communication.

**Comment by fiwi** [comment]
Giving a Commenter / Client the ability to add a to-do would be amazing, the Editor user provides too much information for a client.

**Reply by as1029** [reply]
I would also find the External/Client role a great help - Andy
